### DISCUSSION TITLE
Using file globs as arguments to a typer cli behaves differently on windows and linux

### DISCUSSION THREAD

**Discussion by per11235813** [fastapi/typer#562] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import sys
import typer

print("Data from sys.argv:")
for i, arg in enumerate(sys.argv):
    print(f"sys.argv[{i}] = {arg}")


app = typer.Typer(no_args_is_help=True, add_completion=False)


@app.command()
def find(flist: list[str]):
    """Find files using globs"""
    print("\nFiles found:")
    for f in flist:
        print("  ", f)  


if __name__ == "__main__":
    app()
```

### Description

I am frustrated about performance of find.exe that comes with git-scm, so I want to use python and typer instead. 

I use the script above. I want to use globs on the command line, like `python find.py ~/AppData/**/*ditto*.db`. The attached scripts works and produces the desired output, eg (filename in the output is shorted a bit):

```
> python find.py ~/AppData/**/*ditto*.db
Data from sys.argv:
sys.argv[0] = C:\Users\d6bqr00\Documents\repo\system\utils\find.py
sys.argv[1] = ~/AppData/**/*ditto*.db

Files found:
   C:\Users\....\Ditto.db
```

When looking at sys.argv[1] printed out it is clear that the Windows does not expand the glob, but somehow `typer` or some other library identifies the arguments as a glob and expands it to a list of file names.

I expected to get the argument glob from the command line as a string like  `def find(glob: str)`, but then I get all the file names concatenated as a string. 

On linux the OS tries to expand the glob and if there are no matches it parses the glob string in a single entry list.

I understand the behavior on linux, but on Windows it seems a bit magical that that the string I provide as argument is identified as a glob and expanded.

I am sorry if I miss something obvious, but not understanding this drives me nuts :-)

### Operating System

Linux, Windows

### Operating System Details

Linux is WSL on windows 10

### Typer Version

0.6.1

### Python Version

Python 3.10.7

**Comment by ncoghlan** [comment]
This appears to be inherited from the underlying default Windows behaviour of `click`: https://github.com/pallets/click/issues/2359

**Comment by ncoghlan** [comment]
From local experimentation and the later discussions at https://github.com/fastapi/typer/discussions/649 and https://github.com/fastapi/typer/discussions/1076 I worked out how to turn this off:

```python
if __name__ == "__main__":
    app(windows_expand_args=False)
```

The setting is passed to the CLI invocation, rather than to the app constructor. I can kinda see why `click` has the defaults it does, but glob-expansion by default without support for the single-quote and backslash expansion (as POSIX shells provide) is definitely surprising!

Given Typer's additional type information, I wonder if it might make sense to turn off the default `click` level Windows glob expansion, and only add it back for `list[Path]` parameters (although I guess there would be significant backwards compatibility concerns with doing that).

**Reply by michaelHL** [reply]
Thx!
